src/main/java/io/cdap/delta/bigquery/MultiGCSWriter.java

```diff
@@ -250,16 +259,40 @@ private Schema getTargetSchema(Schema schema) {
           name (string)
           _before_id (long)
           _before_name (string)
+
+        If the source used in the pipeline is generating UN-ORDERED events, then for the same example above,
+        staging table would contain following columns:
+
+          _op (string)
+          _batch_id (long)
+          _sequence_num (long)
+          _source_timestamp (long)
+          _row_id (string)
+          id (long)
+          name (string)
      */
     private Schema getStagingSchema(Schema schema) {
-      List<Schema.Field> fields = new ArrayList<>(2 * schema.getFields().size() + 3);
-      fields.add(Schema.Field.of("_op", Schema.of(Schema.Type.STRING)));
-      fields.add(Schema.Field.of("_batch_id", Schema.of(Schema.Type.LONG)));
+      int fieldLength = sourceEventOrdering == SourceProperties.Ordering.ORDERED ?
+        2 * schema.getFields().size() + 3 // (source schema + _before_ columns) + _op, _batch_id, _sequence_num
+        : schema.getFields().size() + 4; // source schema fields + _op, _batch_id, _sequence_num, _row_id
+
+      List<Schema.Field> fields = new ArrayList<>(fieldLength);
+
+      fields.add(Schema.Field.of(Constants.OPERATION, Schema.of(Schema.Type.STRING)));
+      fields.add(Schema.Field.of(Constants.BATCH_ID, Schema.of(Schema.Type.LONG)));
+      fields.add(Schema.Field.of(Constants.SEQUENCE_NUM, Schema.of(Schema.Type.LONG)));
+
+      // add all fields from source schema
       fields.addAll(schema.getFields());
+
+      if (sourceEventOrdering == SourceProperties.Ordering.UN_ORDERED) {
+        // add _row_id and _source_timestamp fields to handle un-ordered events
+        fields.add(Schema.Field.of(Constants.ROW_ID, Schema.of(Schema.Type.STRING)));
+        return Schema.recordOf(schema.getRecordName() + ".staging", fields);
+      }
+
+      // add _before_ fields for ORDERED events only
       for (Schema.Field field : schema.getFields()) {
-        if ("_sequence_num".equals(field.getName())) {
-          continue;
-        }
         String beforeName = "_before_" + field.getName();
         Schema beforeSchema = field.getSchema();
         beforeSchema = beforeSchema.isNullable() ? beforeSchema : Schema.nullableOf(beforeSchema);
```

Comment:
don't quite understand why we need this line ? if it's nullable , that means it's already a UNION type of one type and NULL.

Reply:
This is to make any field in the staging schema of type `_before_<fieldname>` nullable. So for example if `address` field in the input schema is already nullable then we dont need to take any additional changes however say `id` field in the input schema is not nullable we need to make `_before_id` nullable by calling `Schema.nullableOf(beforeSchema)Schema.nullableOf(beforeSchema)`

Reply:
ah, I see, thanks for explanation :)